File: example-jv-06db/src/main/java/foo/bar/example/foredb/OG.java
package foo.bar.example.foredb;

import android.app.Application;

import java.util.HashMap;
import java.util.Map;

import co.early.fore.core.WorkMode;
import co.early.fore.core.logging.AndroidLogger;
import co.early.fore.core.logging.Logger;
import co.early.fore.core.time.SystemTimeWrapper;
import co.early.fore.net.retrofit2.CallProcessorRetrofit2;
import co.early.fore.net.InterceptorLogging;
import foo.bar.example.foredb.api.CustomGlobalErrorHandler;
import foo.bar.example.foredb.api.CustomGlobalRequestInterceptor;
import foo.bar.example.foredb.api.CustomRetrofitBuilder;
import foo.bar.example.foredb.api.todoitems.TodoItemService;
import foo.bar.example.foredb.db.todoitems.TodoItemDatabase;
import foo.bar.example.foredb.feature.bossmode.BossMode;
import foo.bar.example.foredb.feature.remote.RemoteWorker;
import foo.bar.example.foredb.feature.todoitems.TodoListModel;
import foo.bar.example.foredb.message.UserMessage;
import retrofit2.Retrofit;

import static co.early.fore.core.Affirm.notNull;

/**
 *
 * OG - Object Graph, pure DI implementation
 *
 * Copyright © 2019 early.co. All rights reserved.
 */
public class OG {

    private static boolean initialized = false;
    private static final Map<Class<?>, Object> dependencies = new HashMap<>();


    public static void setApplication(Application application) {
        setApplication(application, WorkMode.ASYNCHRONOUS);
    }

    public static void setApplication(Application application, final WorkMode workMode) {

        notNull(application);
        notNull(workMode);


        // create dependency graph
        AndroidLogger logger = new AndroidLogger("fore_");
        SystemTimeWrapper systemTimeWrapper = new SystemTimeWrapper();
        TodoItemDatabase todoItemDatabase = TodoItemDatabase.getInstance(
                application,
                false,
                workMode);
        TodoListModel todoListModel = new TodoListModel(
                todoItemDatabase,
                logger,
                systemTimeWrapper,
                workMode);
        BossMode bossMode = new BossMode(
                todoListModel,
                systemTimeWrapper,
                workMode,
                logger);

        Retrofit retrofit = CustomRetrofitBuilder.create(
                new CustomGlobalRequestInterceptor(logger),
                new InterceptorLogging(logger));//logging interceptor should be the last one
        CallProcessorRetrofit2<UserMessage> callProcessor = new CallProcessorRetrofit2<UserMessage>(
                new CustomGlobalErrorHandler(logger),
                logger);

        RemoteWorker remoteWorker = new RemoteWorker(
                todoListModel,
                retrofit.create(TodoItemService.class),
                callProcessor,
                systemTimeWrapper,
                logger,
                workMode);


        // add models to the dependencies map if you will need them later
        dependencies.put(BossMode.class, bossMode);
        dependencies.put(TodoListModel.class, todoListModel);
        dependencies.put(RemoteWorker.class, remoteWorker);
        dependencies.put(Logger.class, logger);

    }

    public static void init() {
        if (!initialized) {
            initialized = true;

            // run any necessary initialization code once object graph has been created here
            get(TodoListModel.class).fetchLatestFromDb();
        }
    }

    public static <T> T get(Class<T> model) {

        notNull(model);
        T t = model.cast(dependencies.get(model));
        notNull(t);

        return t;
    }

    public static <T> void putMock(Class<T> clazz, T object) {

        notNull(clazz);
        notNull(object);

        dependencies.put(clazz, object);
    }

}

File: app-examples/example-kt-04retrofit/src/main/java/foo/bar/example/foreretrofitkt/OG.kt
package foo.bar.example.foreretrofitkt

import android.app.Application
import co.early.fore.kt.core.logging.AndroidLogger
import co.early.fore.kt.core.logging.SilentLogger
import co.early.fore.kt.net.InterceptorLogging
import co.early.fore.kt.net.retrofit2.CallWrapperRetrofit2
import foo.bar.example.foreretrofitkt.api.CustomGlobalErrorHandler
import foo.bar.example.foreretrofitkt.api.CustomGlobalRequestInterceptor
import foo.bar.example.foreretrofitkt.api.CustomRetrofitBuilder
import foo.bar.example.foreretrofitkt.api.fruits.FruitService
import foo.bar.example.foreretrofitkt.feature.fruit.FruitFetcher
import java.util.*


/**
 *
 * OG - Object Graph, pure DI implementation
 *
 * Copyright © 2019 early.co. All rights reserved.
 */
@Suppress("UNUSED_PARAMETER")
object OG {

    private var initialized = false
    private val dependencies = HashMap<Class<*>, Any>()

    fun setApplication(application: Application) {

        // create dependency graph

        val logger = if (BuildConfig.DEBUG) AndroidLogger("fore_") else SilentLogger()

        // networking classes common to all models
        val retrofit = CustomRetrofitBuilder.create(
            CustomGlobalRequestInterceptor(logger),
            InterceptorLogging(logger)
        )//logging interceptor should be the last one

        val callWrapper = CallWrapperRetrofit2(
            errorHandler = CustomGlobalErrorHandler(logger),
            logger = logger
        )

        // models
        val fruitFetcher = FruitFetcher(
            retrofit.create(FruitService::class.java),
            callWrapper,
            logger
        )

        // add models to the dependencies map if you will need them later
        dependencies[FruitFetcher::class.java] = fruitFetcher
    }


    fun init() {
        if (!initialized) {
            initialized = true

            // run any necessary initialization code once object graph has been created here

        }
    }


    /**
     * This is how dependencies get injected, typically an Activity/Fragment/View will call this
     * during the onCreate()/onCreateView()/onFinishInflate() method respectively for each of the
     * dependencies it needs.
     *
     * Can use a DI library for similar behaviour using annotations
     *
     * Will return mocks if they have been set previously in putMock()
     *
     *
     * Call it like this:
     *
     * <code>
     *     yourModel = OG[YourModel::class.java]
     * </code>
     *
     * If you want to more tightly scoped object, one way is to pass a factory class here and create
     * an instance where you need it
     *
     */
    @Suppress("UNCHECKED_CAST")
    operator fun <T> get(model: Class<T>): T = dependencies[model] as T

    fun <T> putMock(clazz: Class<T>, instance: T) {
        dependencies[clazz] = instance as Any
    }
}
